src/c2pa.cpp

```diff
@@ -0,0 +1,744 @@
+        case SEEK_CUR:
+            dir = ios_base::cur;
+            break;
+        case SEEK_END:
+            dir = ios_base::end;
+            break;
+        };
+
+        ostream->clear(); // Clear any error flags
+        ostream->seekp(offset, dir);
+        if (ostream->fail())
+        {
+            errno = EINVAL; // Invalid argument
+            return -1;
+        }
+        else if (ostream->bad())
+        {
+            errno = EIO; // Input/output error
+            return -1;
+        }
+        int pos = ostream->tellp();
+        if (pos < 0)
+        {
+            errno = EIO; // Input/output error
+            return -1;
+        }
+        return pos;
+    }
+
+    int CppOStream::writer(StreamContext *context, const void *buffer, int size)
+    {
+        std::ostream *ofstream = (std::ostream *)context;
+        // printf("writer ofstream = %p\n", ofstream);
+        ofstream->write((const char *)buffer, size);
+        if (ofstream->fail())
+        {
+            errno = EINVAL; // Invalid argument
+            return -1;
+        }
+        else if (ofstream->bad())
+        {
+            errno = EIO; // I/O error
+            return -1;
+        }
+        return size;
+    }
+
+    int CppOStream::flusher(StreamContext *context)
+    {
+        std::ofstream *ofstream = (std::ofstream *)context;
+        ofstream->flush();
+        return 0;
+    }
+
+    /// IOStream Class wrapper for CStream implementation.
+    template <typename IOStream>
+    CppIOStream::CppIOStream(IOStream &iostream)
+    {
+        static_assert(std::is_base_of<std::iostream, IOStream>::value, "Stream must be derived from std::iostream");
+        c_stream = c2pa_create_stream(reinterpret_cast<StreamContext *>(&iostream), (ReadCallback)reader, (SeekCallback)seeker, (WriteCallback)writer, (FlushCallback)flusher);
+    }
+    CppIOStream::~CppIOStream()
+    {
+        c2pa_release_stream(c_stream);
+    }
+
+    size_t CppIOStream::reader(StreamContext *context, void *buffer, size_t size)
+    {
+        std::iostream *iostream = (std::iostream *)context;
+        iostream->read((char *)buffer, size);
+        if (iostream->fail())
+        {
+            if (!iostream->eof())
+            {                   // do not report eof as an error
+                errno = EINVAL; // Invalid argument
+                // std::perror("Error: Invalid argument");
+                return -1;
+            }
+        }
+        if (iostream->bad())
+        {
+            errno = EIO; // Input/output error
+            // std::perror("Error: Input/output error");
+            return -1;
+        }
+        size_t gcount = iostream->gcount();
+        return gcount;
+    }
+
+    long CppIOStream::seeker(StreamContext *context, long int offset, int whence)
+    {
+        iostream *iostream = (std::iostream *)context;
+
+        if (!iostream)
+        {
+            errno = EINVAL; // Invalid argument
+            return -1;
+        }
```

Comment:
We could replace this with an `assert(iostream)` so this is apparent to the developer that you aren't able to cast from the StreamContext since that error would be happening at dev time mostly.

Reply:
This is a callbak from Rust code via a C binding, so the assert is unlikely to do the right thing.